horovod/tensorflow/mpi_ops.py

```diff
@@ -99,7 +111,25 @@ def allgather(tensor, name=None):
     return MPI_LIB.horovod_allgather(tensor, name=name)
 
 
-ops.NotDifferentiable('HorovodAllgather')
+@ops.RegisterGradient('HorovodAllgather')
+def allgather_grad(op, grad):
+    """Gradient for allgather op.
+
+    Args:
+      op: An operation.
+      grad: `Tensor` gradient with respect to the output of the op.
+
+    Returns:
+      The gradient with respect to the input of the op.
+    """
+    grad = _allreduce(grad)
+
+    x = op.inputs[0]
+    d = tf.convert_to_tensor([x.shape[0]], dtype=tf.int32)
+    d = tf.reshape(allgather(d), [size()])
+
+    splits = tf.split(grad, num_or_size_splits=d, axis=0)
```

Comment:
I wonder if there's more optimized version of this?

Reply:
I've been wondering that, too.  I think the most expensive part is that we do both an allreduce and allgather here, when it seems we could condense it to a single network call.  But with the current API, we don't have any way (that I know of) to do both at once.

I think in general it will not be too much of an issue, however.  Even in batch normalization, I would expect the user to primarily be using allreduce, not allgather, so I don't currently know of a usecase for this gradient.